esp/esdllib/esdl_script.cpp

```diff
@@ -1671,6 +1671,78 @@ class CEsdlTransformOperationTrace : public CEsdlTransformOperationWithoutChildr
         }
         return false;
     }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) = 0;
+};
+
+class CEsdlTransformOperationTraceContent : public CEsdlTransformOperationTrace
+{
+protected:
+    StringAttr m_contentType;
+    Owned<ICompiledXpath> m_skipMask;
+
+public:
+    CEsdlTransformOperationTraceContent(IXmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationTrace(xpp, stag, prefix)
+    {
+        m_contentType.set(stag.getValue("content_type"));
+        const char* skipMask = stag.getValue("skip_mask");
+        if (!isEmptyString(skipMask))
+            m_skipMask.setown(compileXpath(skipMask));
+    }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) override
+    {
+        bool isValue = false;
+        if (!targetContext->selectText(m_select, text, isValue) || !masker)
+            return true;
+        if (!m_skipMask || !sourceContext->evaluateAsBoolean(m_skipMask))
+        {
+            char* buffer = const_cast<char*>(text.str());
+            masker->maskContent(m_contentType.get(), buffer, 0, text.length());
+        }
+        return true;
+    }
+};
+
+class CEsdlTransformOperationTraceValue : public CEsdlTransformOperationTrace
+{
+protected:
+    XPathLiteralUnion m_valueType;
+    XPathLiteralUnion m_maskStyle;
+
+public:
+    CEsdlTransformOperationTraceValue(IXmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationTrace(xpp, stag, prefix)
+    {
+        m_valueType.setRequired(stag, "value_type", *this);
+        m_maskStyle.setOptional(stag, "mask_style");
+    }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) override
+    {
+        bool isValue = true;
+        if (!targetContext->selectText(m_select, text, isValue))
+        {
+            if (!isValue)
+            {
+                recordWarning(ESDL_SCRIPT_Warning, "ambiguous value xpath");
```

Comment:
Should the xpath that turned out to be ambiguous be output in the warning message to make it easier to track down?

Reply:
Updated warning messages for both warning scenarios, and added tests to confirm that both scenarios do block trace output.